### ISSUE TITLE
Friendly Markdown

### ISSUE THREAD

**Issue by shamrin** [lektor/lektor#7]
Markdown is nice, but it's not so easy for nontechnical editors that Lektor caters to. 

Embedding ProseMirror could help with this problem:

> Say you have a site that allows users to enter comments, and you've decided to use Markdown for the comment input. Your target group mostly knows how to use Markdown, and finds it convenient to type. But you also have some non-technical users, for whom learning arcane syntactic rules does not come naturally.
> 
> Without changing anything in your backend, you can drop in ProseMirror as an alternative input editor. People can even switch between both views as they are editing!

http://prosemirror.net/demo_markdown.html

I hope someone will like this idea so much that I won't need to do it myself :)

**Comment by mitsuhiko** [comment]
Yeah, absolutely. The idea is definitely to have the markdown widget in the admin be one that is some form of WYSIWYG that generates out markdown.  Main reason it's not implemented is because there was not enough time yet and I did not find a good embeddable component.

**Comment by geekpete** [comment]
Prosemirror looks awesome.

**Comment by keybits** [comment]
+1 for ProseMirror - would be ideal here

**Comment by KarimJedda** [comment]
Hey, I'm using https://github.com/sofish/pen as a text input editor for Lektor, it has markdown support and looks pretty nice :+1: I think it could be easier to integrate. Anyways, if someone is working towards this I would love to help (I know nothing about react though. . .)

**Comment by fmartingr** [comment]
Came here to propose [stackedit](https://stackedit.io/editor) as an alternative but [pen](https://github.com/sofish/pen) seems better for nontechnical users.

**Comment by tylerlong** [comment]
I recommend [Markdown Plus](http://mdp.tylingsoft.com/). It is open source under MIT license. I am the author and I am willing to provide any assistance necessary.

**Comment by ghost** [comment]
Be present to implement [Common Markdown](http://commonmark.org/)

**Comment by kmonsoor** [comment]
@tylerlong and others, if someone can make a basic integration of `Lektor` and any of these good editors(`Markdown plus`, `sofish/pen` etc.) that would be a great starting point.

**Comment by karelv** [comment]
+1 for Markdown Plus

**Comment by jungornti** [comment]
+1 for Stackedit

**Comment by mfeif** [comment]
+1 for a basic integration example / fork :-)

**Comment by fmartingr** [comment]
I've tried PEN on another project and there's a thing to keep in mind: It's **not** a Markdown editor, it's a WYSIWYG editor (via `contentEditable`) with markdown "macros" and that exports to markdown.

It may seem the same, but at least for me it ain't.

**Comment by vassudanagunta** [comment]
Perhaps @abnerlee will contribute Typora, the best WYSIWYG Markdown editor I've tried yet.

**Comment by rhysmoyne** [comment]
Hi all,

I had a go at integrating SimpleMDE with Lektor. This was mainly for the purposes of administering my   own website I am creating. Nonetheless, I thought it would be a good idea to share what I did here, even if SimpleMDE is not the editor you go with! 

Screenshot:

![2016-09-22-091940_1680x1050_scrot](https://cloud.githubusercontent.com/assets/5418753/18732986/6c2ef1e6-80a6-11e6-9df8-4d89d274978a.png)

Please see https://github.com/lektor/lektor/compare/master...rhysmoyne:master

Things to do:

I have configured SimpleMDE with controls suitable for my project (see line 237 of primitiveWidgets.jsx), need to choose a suitable set of default buttons for all Lektor projects (or should this be configurable?)

This only works on one Markdown textarea at the moment (when you create a new SimpleMDE object it assigns itself to the first textarea it finds but this can be changed).

Any other feedback or suggestions from those more familiar with the Lektor codebase is welcome!

Rhys.

**Comment by nixjdm** [comment]
I'm marking this as on-hold pending the outcome of #457. This may well be rolled into that ticket.

**Comment by ericmjl** [comment]
@nixjdm What's the status now on this issue?

**Comment by nixjdm** [comment]
Since we've had further discussion on #457 / #458, I'm removing the on-hold label. Though, it still may be prudent to handle #458 first, but I won't stop someone from picking this up.

**Comment by davidferguson** [comment]
Having a GUI Markdown editor is something that I need in Lektor, and due to this I've gone and created a widget for [TOAST UI Editor](https://github.com/nhnent/tui.editor) which I've added to [my fork](https://github.com/davidferguson/lektor/tree/friendly-markdown) of Lektor (with a PR coming soon, once I've added a couple more things).

TOAST UI Editor was chosen as it:

* Is a full WYSIWYG Markdown editor
* Implements Commonmark and GFM standards
* Allows switching between Rich Text (WYSIWYG) mode and raw Markdown mode, and changes are reflected between modes
* Has the option of adding extensions, which may be useful in the future
* Allows pasing of copied Rich Text and HTML directly into the editor whilst keeping (most of) the formatting

Comparing this to the other editors mentioned in this thread:

* **Prosemirror**: Does not implement Commonmark or GFM, so this would need to be manually added
* **Pen**: Is a form of hybrid Markdown-GUI editor, doesn't show raw Markdown text
* **StackEdit**: Only has a GUI preview mode, not edit mode so all edits must be done in Markdown
* **Markdown Pus**: Same as StackEdit
* **Typora**: Appears to be a desktop app, so not relevant here

- - -

Differing behaviour:

* The `size` [type option](https://www.getlektor.com/docs/models/) has no effect on the GUI editor, as there is no API in TUI-Editor to scale the text
* The Markdown GUI Editor doesn't resize as text is entered, like `multiline-text` does

It should be noted that I've used the existing markdown type (from `formats.py`), which means that when you enter text in the GUI, it's saved as Markdown and then converted to HTML using `markdown.py`. However the GUI editor allows you to directly get the HTML, and there might be a case that this is better than using the existing method, as it means the rendered HTML will be exactly as what's shown in the GUI preview. Thoughts on this would be appreciated.

Something else to consider is that the GUI editor [claims to respect Commonmark and GFM](https://github.com/nhnent/tui.editor#commonmark--gfm-specifications) (with optional extension) and will possibly produce undesired results when given Markdown that doesn't conform to this. This could cause issues for some pre-existing installations, and so thoughts on this would be useful. Could it be made that only new installs get the GUI Markdown editor as the default, as to not disrupt existing installations? Or maybe just make a new 'markdown-gui' type, so people can still use the old 'markdown' type if they prefer?

**Comment by runfalk** [comment]
Great work! I think lack of a proper Markdown editor has been a pain point of Lektor for a really long time. I haven't tried your fork yet but I'll share some thoughts that come to mind.

I think ignoring `size` is fine. I think the main purpose of the attribute is for single line inputs anyway.

I think not resizing dynamically is bad. If possible I'd like that to be added. It's fine for models with just a few attribute fields and a main text area. But for flowblocks there will be so many editors with their own scroll and fixed size that it will be harder to find the group of text that you want.

Another thing I think is important is that the preview should not be a split view like in the examples. I think it should work like the GitHub preview does with a separate tab. I also think the Markdown tab should be the default, but that may just be my programmer mindset.

I think saving in Markdown is a good thing despite the possible inconsistency. The preview will not have access to the correct stylesheet anyway, so it's not a substitute for the default preview anyway. Saving to HTML may garble the Markdown, if I write something in Markdown and save I should have the exact same Markdown when I reload.

Following the same logic I don't think Commonmark/GFM conformance should be an issue either. Lektor uses [mistune](https://github.com/lepture/mistune), and I don't  know which standard (if any) it conforms to. As long as existing installations are unaffected I'm happy.

I'd wait with making the changes I suggested until other maintainers have spoken their minds. After all I can't speak for everyone.

**Comment by davidferguson** [comment]
>I think not resizing dynamically is bad.

I'll have a look into whether it's possible to do.

>Another thing I think is important is that the preview should not be a split view like in the examples.

Maybe I didn't explain this well enough - it *is* tab based (with an optional preview split view on the Markdown tab which I've disabled as it looked too cramped when embedded with Lektor, and isn't really necessary when you can switch to the WYSIWYG view to preview it)

**Comment by nixjdm** [comment]
Thank you @davidferguson! I will try this out soon too. Here are my immediate thoughts.

It is important that this not change how existing sites are rendered. Ideally it can load markdown content, toggle WYSIWYG on/off, and still save the same markdown. At least loading and saving should preserve the same content. As @runfalk pointed out, Lektor will do the final rendering in Python, and we shouldn't make an assumption how that will happen. Especially since there are markdown events that can be hooked to change things, and even [at least one different renderer](https://www.getlektor.com/plugins/lektor-pythonmarkdown/) that can be used.

I agree that I very much would like this field to dynamically resize. I also agree that the split view should default off, though I wonder if it would make more sense if we allowed the admin to be full screen width.

The only drawback of this I can think of is not rendering with the renderer Lektor uses. Most of the time I think people won't spot a difference, unless they've changed from the default mistune. I think that's fine, but I expect we may get an occasional bug report because things don't match. There's probably not a lot that can be done about this generally though.

**Comment by davidferguson** [comment]
>I agree that I very much would like this field to dynamically resize.

I've just added dynamically resizing into my fork now, for both the wysiwyg and markdown editors.

>I also think the Markdown tab should be the default, but that may just be my programmer mindset.

Currently I've got the WYSIWYG editor as default, as I think this will be easier for non-technical users, and people who know what markdown is can just switch to that tab.

**Comment by nixjdm** [comment]
@davidferguson Sorry if you've been waiting on me. I've tried this out, and I love it. I didn't have any problems using it, tested on node 6 and 11.

The only quibble I have left is a very minor style opinion. By default, since we have Rich Text selected by default, I'd prefer it be on the left. That seems more intuitive to me (at least for left-to-right language users). ![wyswyg](https://user-images.githubusercontent.com/3431410/56048703-aa9ec280-5d0d-11e9-8728-8492c03c3743.png)

Edit: It also might be appreciated if we allowed a project file variable to set the default to Markdown instead of Rich Text project-wide. I think that also could be an additional ticket, not required for a PR.

Even without that change though, this looks really good and was a joy to use. Do you or @runfalk have any additional concerns about this?

**Comment by runfalk** [comment]
@davidferguson I tried out your fork and I like it, great work! I tested it on Node 8.10 with Python 2.7.

I've found one bug. I can't go to the Edit page and then click Preview (or leave the page) without a prompt about unsaved changes appearing. It seems like the change detection rules don't work. This needs to be fixed before we can merge this.

There are also few things I would like to alter unless it's a lot of work for you:

* The existing fields use the Roboto Slab font (I think). I'd like at least the markdown view to use that, for visual consistency.
* I also think it's a bit weird to have the switcher on the bottom. If possible I think it fits better in the toolbar like the Write/Preview tab when you reply to GitHub issues.
* If you go between the Edit and Preview pages it loses track of which mode that was selected.
* Personally, I'll never use the Rich text view. I would like it if there was some way of toggling the default mode to use for the field. I think it's acceptable to configure this in the model files.
* I think the visuals for the Rich Text rendering could be improved a lot and match closer with the overall Lektor aesthetics, though this should probably be done tracked as a separate issue.

**Comment by davidferguson** [comment]
@nixjdm @runfalk Most of those issues/comments raised have now been addressed.

>The only quibble I have left is a very minor style opinion. By default, since we have Rich Text selected by default, I'd prefer it be on the left.

The mode switch buttons have been moved from the bottom of the widget to the top toolbar, and Rich Text is now on the right.

---

>Edit: It also might be appreciated if we allowed a project file variable to set the default to Markdown instead of Rich Text project-wide. I think that also could be an additional ticket, not required for a PR.

>Personally, I'll never use the Rich text view. I would like it if there was some way of toggling the default mode to use for the field. I think it's acceptable to configure this in the model files.

I've created a new type `markdown-gui` (and made the old `markdown` type use the old textarea editor again) which has the option for specifying the default view in the modal file. If an unknown value is used or it is left unspecified, it defaults to the Rich Text editor (which aligns with the Rich Text tab being on the right).

Having this implemented as a separate type means that people who have been using the old `markdown` type won't see anything different. There are positives and negatives to this, but imo I think having a new type (which would be the default in the example project) is the better option.

**Note: because of the change to `markdown-gui`, when testing in my fork you'll need to change the type of one of the fields in one of the model files from `markdown` to `markdown-gui` otherwise it'll look like nothing is happening**

---

>I've found one bug. I can't go to the Edit page and then click Preview (or leave the page) without a prompt about unsaved changes appearing. It seems like the change detection rules don't work. This needs to be fixed before we can merge this.

Fixed.

---

>The existing fields use the Roboto Slab font (I think). I'd like at least the markdown view to use that, for visual consistency.

The Markdown tab of the editor now uses Roboto Slab (with the same backup fonts as the old textarea input). I've left the Rich Text tab font alone, as to me it looks acceptable the way it is, and shouldn't use a monospaced font.

---

>If you go between the Edit and Preview pages it loses track of which mode that was selected.

This is due to how the Edit and Preview views are rendered by React. When you switch between them it's effectively initialising them from the start, so there's not an easy way to retain what view the editor was in. This *could* be fixed by saving the editor view state to the parent react component (which I believe is `main.jsx`), but I think having the ability to specify the default view in the modal file reduces the need for this.

---

>I think the visuals for the Rich Text rendering could be improved a lot and match closer with the overall Lektor aesthetics, though this should probably be done tracked as a separate issue.

The editor is fairly customisable, so this should be possible. I've done a little bit (made the mode switch buttons Lektor-y, and added two widgets - see additional comment below) but indeed this can be improved, and should probably be done in a separate issue as suggested.

**Comment by davidferguson** [comment]
Aside from implementing those fixes/improvements, I've also replaced the standard editor image and link widgets with custom ones that better integrate with the Lektor system.

The Image widget has three tabs:
- **Attachment**: Allows selecting of existing image attachment to add into the editor.
- **Upload**: Allows uploading a new attachment to that page, which will then be inserted into the editor.
- **External**: Simple external image URL.

The link widget has two tabs:
- **Page**: Allows searching for pages to link to (using modified version of the existing search dialog)
- **URL**: Simple external href link.

I've tried to re-use as much existing widgets/code as possible for these, but it's not always been possible due to the structure of Lektor, and the way the editor and plugins work. Hopefully they're acceptable though.

Below are screenshots of the two widgets:

![image-widget](https://user-images.githubusercontent.com/13520342/57048744-2c2ea400-6c6d-11e9-9d57-b05d0a9ae102.png)

![link-widget](https://user-images.githubusercontent.com/13520342/57048764-3b155680-6c6d-11e9-83bc-b3703150fcf3.png)

If there are no more suggestions/fixes needed, I'll go ahead and open this as a PR.

**Comment by runfalk** [comment]
Great work @davidferguson. I tried it out on a project of mine and I like the aesthetic a lot more. I like the `default_view` option as well. I think it alleviates the need for remembering the tab state when switching between preview and edit.

I think I agree with you choice of not changing the rich text font, as sans-serif is pretty much the de-facto default for the web.

Based on your screenshots I like what you did with the link and image widgets. However, I was unable to get it to work on my end:

* Firefox 66/Chromium 73
* Python 3.7.3
* Node 10.15.3
* NPM 6.4.1

I checked out your branch in a clean repository and did `make build-js`.

The error I'm getting is: `TypeError: this.props.params is undefined` at `RecordComponent.jsx:25:10`. It appears every time I type in the page search box or try to click the *Add image* button.

Looking forward to review your PR :smile:

**Comment by davidferguson** [comment]
Hmm, odd. I've just tried this on a different machine and did a fresh clone using:

```
git clone https://github.com/davidferguson/lektor
cd lektor
git checkout friendly-markdown
virtualenv --python=/usr/bin/python3 venv
source venv/bin/activate
pip install --editable .
make build-js
cp -r example example-project
nano example-project/models/page.ini # change type from markdown to markdown-gui
lektor --project example-project server
```

Does that align with what you did?

Using:
- Python 3.7.3
- npm 6.4.1
- node v10.15.3
- Chrome 72.0

**Comment by runfalk** [comment]
I didn't do this specifically, I tried it on my own site. The setup steps look really similar though. Maybe it has to do with me using the editor inside flowblocks. I'll try to follow your steps when I get home.

**Comment by davidferguson** [comment]
Ah, that could be it. I've never tested with flowblocks, and thinking on how I did it, that could well be the case. I'll investigate.

**Comment by davidferguson** [comment]
Yep my apologies @runfalk - I forgot to add in a bit to make it work in flow blocks. It's added in and I've found it to be working in my tests now.

**Comment by runfalk** [comment]
@davidferguson I tried it out, and now it works for me with the flowblocks. I think you should open a PR now so we can review the code properly.

Great work!

**Comment by apatrushev** [comment]
This issue is close to five years of age, 18 people have contributed, there is good discussion extending to important issues such as which font to use. Still Lektor does not come with WYSIWYG editor experience, something that all content managers except us nerds have come to expect. I wrote a plugin that does just that: adds SimpleMDE editor for all markdown fields, in case you have come to this issue looking for a solution.

https://github.com/sphericalpm/lektor-simplemde
https://pypi.org/project/lektor-simplemde/

**Comment by davidferguson** [comment]
My PR should be considered dead now (unless someone else wants to pick it up, of course). It did work okay, but there were slight bugs, and the project I was using lektor for adopted a custom markup language rather than markdown, which is why I never continued it.
